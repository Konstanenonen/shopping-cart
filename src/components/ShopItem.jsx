import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styles from '../styles/components/ShopItem.module.scss';
import Button from './Button';

function ShopItem({ title, description, addToCart }) {
  const [amount, setAmount] = useState(0);
  const amountInputId = 'item-amount';

  const handleClick = () => {
    addToCart(amount);
    setAmount(0);
  };

  return (
    <div className={styles.container}>
      <h3 className={styles.title}>{title}</h3>
      <p>{description}</p>
      <div className={styles.row}>
        <Button handleClick={() => setAmount(amount - 1)}>-1</Button>
        <label className={styles.label} htmlFor={amountInputId}>
          Add more to order:
          <input
            value={amount}
            onChange={(e) => setAmount(Number(e.target.value))}
            type="number"
            id={amountInputId}
          />
        </label>
        <Button handleClick={() => setAmount(amount + 1)}>+1</Button>
      </div>
      <Button handleClick={handleClick}>Add to cart</Button>
    </div>
  );
}

ShopItem.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  addToCart: PropTypes.func.isRequired,
};

export default ShopItem;
