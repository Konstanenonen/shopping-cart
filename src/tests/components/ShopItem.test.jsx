/* global it describe expect jest */

import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ShopItem from '../../components/ShopItem';

describe('ShopItem component', () => {
  it('renders "TITLE" and "DESCRIPTION"', () => {
    const { container } = render(
      <ShopItem title="TITLE" description="DESCRIPTION" addToCart={() => {}} />
    );
    expect(container).toMatchSnapshot();
  });

  it('calls addToCart correct amount of times', () => {
    const addToCartMock = jest.fn();
    render(
      <ShopItem
        title="TITLE"
        description="DESCRIPTION"
        addToCart={addToCartMock}
      />
    );
    const button = screen.getByRole('button', { name: 'Add to cart' });

    userEvent.click(button);
    userEvent.click(button);
    userEvent.click(button);

    expect(addToCartMock).toBeCalledTimes(3);
  });
});
