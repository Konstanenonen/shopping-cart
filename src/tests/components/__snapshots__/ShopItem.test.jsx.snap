// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopItem component renders "TITLE" and "DESCRIPTION" 1`] = `
<div>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      TITLE
    </h3>
    <p>
      DESCRIPTION
    </p>
    <div
      class="row"
    >
      <button
        class="button"
        type="button"
      >
        -1
      </button>
      <label
        class="label"
        for="item-amount"
      >
        Add more to order:
        <input
          id="item-amount"
          type="number"
          value="0"
        />
      </label>
      <button
        class="button"
        type="button"
      >
        +1
      </button>
    </div>
    <button
      class="button"
      type="button"
    >
      Add to cart
    </button>
  </div>
</div>
`;
